docs/design/Extension_Management.md

```diff
@@ -0,0 +1,151 @@
+# Extension Management
+
+Upgrade Assistant's extensibility provides the ability for users to extend it beyond its original design. As we've worked with developers making extensions, it's clear that there are some issues with discoverability and maintainability of custom extensions. Some of the major pain points:
+
+- **Version management:** Need to be able to specify which extension version a project should use
+- **Distribution** Need to be able easily get a specific version (or update existing ones) of an extension
+- **Consistency between runs** Should be able to register an extension so subsequent runs automatically use it
+
+Throughout this document, the term `workspace` will be used to refer to a solution or project that is being updated along with the state file Upgrade Assistant generates.
+
+## User Experience
+
+Extensions will now be something that are managed at a workspace level. This will allow users to have a consistent experience as well as be able to manage their extensions.
+
+Below are the commands to manage extensions:
+
+```
+upgrade-assistant extension list
+```
+
+This command will list current extensions applied to a project.
+
+```
+upgrade-assistant extension add --name [name] [--source [source]]
+```
+
+This command will allow users to add extensions to a project.
+
+```
+upgrade-assistant extension remove --name [name]
+```
+
+This command will remove an installed extension from a current workspace.
+
+```
+upgrade-assistant extension update [--name [name]]
+```
+
+This command will update all extensions, or a specific one if the name is given.
+
+```
+upgrade-assistant extension restore
+```
+
+This command will restore extensions when they are not available
+
+> This may be able to be done automatically, but initially, we'll have it be a separate command
```

Comment:
More as it's a check we'd have to initially run that seems potentially expensive. I'm fine though just doing it automatically and seeing how that works.

Reply:
Either way. I expect if an extension is installed, users will want to use it. And we should only have to do that once since we'll cache extensions locally, so doing it automatically seems ok to me.